File: frontend/src/lib/localGame.ts
import { HAS_TAURI, invoke } from '@/lib/tauri'

export type LocalGameRoundView = {
  roundLabel: string
  honba: number
  kyotaku: number
  remainingTiles: number
  dealerSeat: 0 | 1 | 2 | 3
}

export type LocalGamePlayerView = {
  seat: 0 | 1 | 2 | 3
  relationLabel: string
  wind: 'E' | 'S' | 'W' | 'N'
  score: number
  riverTiles: string[]
  melds: string[][]
  statusTags: string[]
  isDealer: boolean
  isSelf: boolean
}

export type LocalGameActionView = {
  id: number
  type: string
  label: string
  hint: string
  enabled: boolean
  tile?: string | null
  mjai?: Record<string, unknown> | null
}

export type LocalGameRecommendationView = {
  rank: number
  actionId: number | null
  tile: string | null
  label: string
  source: string
  status: string
  note: string
}

export type LocalArtifactStatus = {
  saved: boolean
  replayPath: string | null
  decisionPointsPath: string | null
  errorMessage: string | null
}

export type LocalReviewKeyChoice = {
  turnIndex: number
  humanActionLabel: string
  recommendedActionLabel: string | null
  humanTile: string | null
  recommendedTile: string | null
  category: string
}

export type LocalReviewSummary = {
  schemaVersion: number
  source: string
  totalDecisions: number
  top1Matches: number
  mismatchCount: number
  keyChoices: LocalReviewKeyChoice[]
  note: string
}

export type LocalGameView = {
  schemaVersion: number
  source: string
  phaseLabel: string
  notice: string
  round: LocalGameRoundView
  players: LocalGamePlayerView[]
  selfHandTiles: string[]
  doraIndicators: string[]
  actions: LocalGameActionView[]
  recommendations: LocalGameRecommendationView[]
  artifactStatus: LocalArtifactStatus
  reviewSummary: LocalReviewSummary
}

export type LocalGameSessionHandle = {
  gameId: string
  view: LocalGameView
}

export type LocalGameLoadResult = {
  gameId: string
  view: LocalGameView
  mode: 'tauri' | 'fallback'
  error?: string
}

const DEV_FALLBACK_VIEW: LocalGameView = {
  schemaVersion: 1,
  source: 'frontend_dev_fallback',
  phaseLabel: 'Dev fallback',
  notice: 'Vite browser fallback only. Tauri desktop uses local_game_new.',
  round: {
    roundLabel: 'E1',
    honba: 0,
    kyotaku: 0,
    remainingTiles: 62,
    dealerSeat: 0,
  },
  selfHandTiles: ['123m', '405p', '678s', '11z'],
  doraIndicators: ['5m'],
  players: [
    {
      seat: 0,
      relationLabel: '自家',
      wind: 'E',
      score: 25000,
      riverTiles: ['1m', '2p', '9s'],
      melds: [],
      statusTags: ['thinking'],
      isDealer: true,
      isSelf: true,
    },
    {
      seat: 1,
      relationLabel: '下家',
      wind: 'S',
      score: 25000,
      riverTiles: ['3m', '7p', 'P'],
      melds: [['P', 'P', 'P']],
      statusTags: [],
      isDealer: false,
      isSelf: false,
    },
    {
      seat: 2,
      relationLabel: '对面',
      wind: 'W',
      score: 25000,
      riverTiles: ['9m', '2s', 'C'],
      melds: [],
      statusTags: [],
      isDealer: false,
      isSelf: false,
    },
    {
      seat: 3,
      relationLabel: '上家',
      wind: 'N',
      score: 25000,
      riverTiles: ['4m', '8p', 'F'],
      melds: [['3p', '3p', '3p']],
      statusTags: [],
      isDealer: false,
      isSelf: false,
    },
  ],
  actions: [
    {
      id: 1,
      type: 'discard',
      label: 'Discard 5p',
      hint: 'Dev fallback action',
      enabled: false,
      tile: '5p',
      mjai: { type: 'dahai', actor: 0, pai: '5p' },
    },
    { id: 2, type: 'riichi', label: 'Riichi', hint: 'Not wired yet', enabled: false },
    { id: 3, type: 'skip', label: 'Skip', hint: 'Not wired yet', enabled: false },
  ],
  recommendations: [
    {
      rank: 1,
      actionId: 1,
      tile: '5p',
      label: 'Recommended discard 5p',
      source: 'frontend_dev_fallback',
      status: 'recommended',
      note: 'Dev fallback only; not a real AI recommendation.',
    },
  ],
  artifactStatus: {
    saved: false,
    replayPath: null,
    decisionPointsPath: null,
    errorMessage: null,
  },
  reviewSummary: {
    schemaVersion: 1,
    source: 'frontend_dev_fallback',
    totalDecisions: 0,
    top1Matches: 0,
    mismatchCount: 0,
    keyChoices: [],
    note: 'Dev fallback only; no local review summary.',
  },
}

export async function loadLocalGameView(): Promise<LocalGameLoadResult> {
  if (!HAS_TAURI) {
    return { gameId: 'frontend-dev-fallback', view: DEV_FALLBACK_VIEW, mode: 'fallback' }
  }

  try {
    const session = await invoke<LocalGameSessionHandle>('local_game_new')
    return { gameId: session.gameId, view: session.view, mode: 'tauri' }
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error)
    return {
      gameId: 'frontend-dev-fallback',
      view: {
        ...DEV_FALLBACK_VIEW,
        notice: `Tauri command failed; showing dev fallback. ${message}`,
      },
      mode: 'fallback',
      error: message,
    }
  }
}

export async function getLocalGameView(gameId: string): Promise<LocalGameView> {
  return await invoke<LocalGameView>('local_game_get_view', { gameId })
}

export async function submitLocalGameAction(gameId: string, actionId: number): Promise<LocalGameView> {
  return await invoke<LocalGameView>('local_game_submit_action', { gameId, actionId })
}

File: src/schema/local_game.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LocalGameSessionHandle {
    pub game_id: String,
    pub view: LocalGameView,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LocalGameSubmitActionRequest {
    pub game_id: String,
    pub action_id: u32,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LocalGameView {
    pub schema_version: u32,
    pub source: String,
    pub engine: LocalGameEngineMetadata,
    pub phase_label: String,
    pub notice: String,
    pub round: LocalGameRoundView,
    pub players: Vec<LocalGamePlayerView>,
    pub self_hand_tiles: Vec<String>,
    pub dora_indicators: Vec<String>,
    pub actions: Vec<LocalGameActionView>,
    pub recommendations: Vec<LocalGameRecommendationView>,
    #[serde(default)]
    pub artifact_status: LocalArtifactStatus,
    #[serde(default)]
    pub review_summary: LocalReviewSummary,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LocalGameEngineMetadata {
    pub schema_version: u32,
    pub source: String,
    pub status: String,
    pub capabilities: Vec<String>,
    pub note: String,
    #[serde(default)]
    pub worker: LocalWorkerMetadata,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LocalWorkerMetadata {
    pub schema_version: u32,
    pub configured: bool,
    pub agent_kind: String,
    pub label: String,
    pub timeout_ms: Option<u32>,
}

impl Default for LocalWorkerMetadata {
    fn default() -> Self {
        Self {
            schema_version: 1,
            configured: false,
            agent_kind: "random_agent".into(),
            label: "RandomAgent".into(),
            timeout_ms: None,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LocalArtifactStatus {
    pub saved: bool,
    pub replay_path: Option<String>,
    pub decision_points_path: Option<String>,
    pub error_message: Option<String>,
}

impl LocalArtifactStatus {
    pub fn pending() -> Self {
        Self {
            saved: false,
            replay_path: None,
            decision_points_path: None,
            error_message: None,
        }
    }

    pub fn saved(replay_path: String, decision_points_path: String) -> Self {
        Self {
            saved: true,
            replay_path: Some(replay_path),
            decision_points_path: Some(decision_points_path),
            error_message: None,
        }
    }

    pub fn failed(message: String) -> Self {
        Self {
            saved: false,
            replay_path: None,
            decision_points_path: None,
            error_message: Some(message),
        }
    }
}

impl Default for LocalArtifactStatus {
    fn default() -> Self {
        Self::pending()
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LocalGameRoundView {
    pub round_label: String,
    pub honba: u32,
    pub kyotaku: u32,
    pub remaining_tiles: u32,
    pub dealer_seat: u8,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LocalGamePlayerView {
    pub seat: u8,
    pub relation_label: String,
    pub wind: String,
    pub score: i32,
    pub river_tiles: Vec<String>,
    pub melds: Vec<Vec<String>>,
    pub status_tags: Vec<String>,
    pub is_dealer: bool,
    pub is_self: bool,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LocalGameActionView {
    pub id: u32,
    #[serde(rename = "type")]
    pub action_type: String,
    pub label: String,
    pub hint: String,
    pub enabled: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tile: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub mjai: Option<Value>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LocalGameRecommendationView {
    pub rank: u32,
    #[serde(default)]
    pub action_id: Option<u32>,
    #[serde(default)]
    pub tile: Option<String>,
    pub label: String,
    pub source: String,
    pub status: String,
    pub note: String,
    #[serde(default)]
    pub reason: Option<String>,
    #[serde(default)]
    pub elapsed_ms: Option<f64>,
    #[serde(default)]
    pub meta: Option<Value>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LocalReviewSummary {
    pub schema_version: u32,
    pub source: String,
    pub total_decisions: u32,
    pub top1_matches: u32,
    #[serde(default)]
    pub top3_matches: u32,
    pub mismatch_count: u32,
    #[serde(default)]
    pub attention_count: u32,
    #[serde(default)]
    pub fallback_count: u32,
    #[serde(default)]
    pub unavailable_count: u32,
    #[serde(default)]
    pub not_ranked_count: u32,
    pub key_choices: Vec<LocalReviewKeyChoice>,
    pub note: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LocalReviewKeyChoice {
    pub turn_index: u32,
    pub human_action_label: String,
    pub recommended_action_label: Option<String>,
    pub human_tile: Option<String>,
    pub recommended_tile: Option<String>,
    pub category: String,
    #[serde(default)]
    pub reason: Option<String>,
}
